lib/dns/dns.go

```diff
@@ -639,3 +640,94 @@ func (s *Server) Stop() {
 func (s *Server) Wait() {
 	s.wg.Wait()
 }
+
+type FQDNs map[string]bool
+
+// ReverseLookup returns a set of FQDNs for ipAddr from nameservers in /etc/resolv.conf
+// /etc/hosts and /etc/nsswitch.conf are ignored by this function
+func ReverseLookup(ipAddr string) (domains FQDNs) {
+	domains = make(map[string]bool)
+
+	address, err := reverseaddr(ipAddr)
+	if err != nil {
+		log.Errorf("%s", err)
+		return
+	}
+
+	nameservers := resolvconf()
+	log.Infof("Looking up rDNS record for %s\n", address)
+	for _, n := range nameservers {
+		dnsClient := new(mdns.Client)
+		msg := new(mdns.Msg)
+
+		msg.SetQuestion(address, mdns.TypePTR)
+		r, _, err := dnsClient.Exchange(msg, fmt.Sprintf("%s:53", n))
+		if err != nil {
+			log.Warnf("Got error \"%s\" from %s\n", err, n)
+			continue
+		}
+
+		if len(r.Answer) == 0 {
+			log.Warnf("No reply from %s\n", n)
+			continue
+		}
+
+		for _, a := range r.Answer {
+			switch a := a.(type) {
+			case *mdns.PTR:
+				// cut the . off the end of the returned record & store it
+				domains[strings.TrimSuffix(a.Ptr, ".")] = true
+			default:
+				log.Debugf("Got nonstandard answer %s (from nameserver %s)\n", a, n)
+			}
+		}
+	}
+
+	return
+}
+
+// Convert unsigned integer to decimal string.
+// this helper func was lifted from stdlib -- net/dnsclient.go
+func uitoa(val uint) string {
```

Comment:
Is this not viable? https://golang.org/pkg/strconv/#FormatUint

Reply:
I'm not sure if they're the same. This is the function used by `reverseaddr` in the standard lib, but it's possible that they could've used `strconv`

Reply:
I'll compare the functions and see if there's a difference. I sort of suspect that the format of the unsigned int here is defined according to a DNS RFC rather than Go, but it's possible that the std lib is replicating functionality unnecessarily.